### PR TITLE
Expand ARM local variable R11-based addressing

### PR DESCRIPTION
**Pull request by BruceForstall**
When addressing a local with negative offsets from R11 (if it can't
be reached from SP), allow the full range of negative offsets allowed
in the instructions. Floating-point load/store especially has a much
bigger range than what was previously allowed.

### CODE REVIEW

**Comment by BruceForstall** [comment]
@dotnet-bot test Ubuntu arm Cross Checked jitstress1 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked normal Build and Test
@dotnet-bot test Windows_NT arm Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Windows_NT arm Cross Checked normal Build and Test

**Comment by BruceForstall** [comment]
Diffs:
```
Summary:
(Lower is better)
Total bytes of diff: -1984 (-0.01% of base)
    diff is an improvement.
Top file regressions by size (bytes):
          16 : NuGet.Packaging.dasm (0.01% of base)
           8 : Newtonsoft.Json.dasm (0.00% of base)
           8 : System.Linq.Parallel.dasm (0.00% of base)
           8 : System.Security.Cryptography.Primitives.dasm (0.03% of base)
           8 : System.Transactions.Local.dasm (0.01% of base)
Top file improvements by size (bytes):
       -1884 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm (-0.06% of base)
         -96 : System.Data.Common.dasm (-0.01% of base)
         -16 : System.Linq.dasm (-0.01% of base)
          -8 : NuGet.Common.dasm (-0.03% of base)
          -8 : NuGet.Versioning.dasm (-0.03% of base)
13 total files with size differences (8 improved, 5 regressed), 117 unchanged.
Top method regessions by size (bytes):
          16 : NuGet.Packaging.dasm - NuGet.Packaging.PackageFolderReader:CopyFiles(ref,ref,ref,ref,struct):ref:this
           8 : Newtonsoft.Json.dasm - Newtonsoft.Json.JsonReader:ReadArrayIntoByteArray():ref:this
           8 : System.Linq.Parallel.dasm - System.Linq.Parallel.ParallelEnumerableWrapper`1[Vector`1][System.Numerics.Vector`1[System.Single]]:GetEnumerator():ref:this
           8 : System.Security.Cryptography.Primitives.dasm - <WriteAsyncCore>d__49:MoveNext():this
           8 : System.Transactions.Local.dasm - System.Transactions.VolatileEnlistmentSPC:EnterState(ref):this
Top method improvements by size (bytes):
        -360 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.ApplicationServerTraceEventParser:EnumerateTemplates(ref,ref):this
        -360 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.KernelTraceEventParser:EnumerateTemplates(ref,ref):this
        -300 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.CtfTraceEventSource:InitEventMap():ref
        -288 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.ClrPrivateTraceEventParser:EnumerateTemplates(ref,ref):this
        -288 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.ClrTraceEventParser:EnumerateTemplates(ref,ref):this
18 total methods with size differences (13 improved, 5 regressed), 191280 unchanged.
```
Diffs are as expected: more references using R11.

**Comment by BruceForstall** [comment]
x86 failure is in D:\j\workspace\x86_release_w---80850631\bin\tests\Windows_NT.x86.Release\JIT\HardwareIntrinsics\X86\Fma_Vector256\Fma_ro\Fma_ro.cmd. Since this change is completely under `#ifdef _TARGET_ARM_`, it can be ignored.

**Comment by BruceForstall** [comment]
@briansull PTAL
cc @dotnet/jit-contrib @dotnet/arm32-contrib

File: src/jit/compiler.hpp [hunk 1]
```diff
@@ -2434,17 +2434,19 @@ inline
...
+
+            // Use sp encoding. During encoding, we'll pick the best encoding for the actual offset we have.
+            if (opts.MinOpts() || (actualOffset <= encodingLimitUpper))
             {
                 offset    = spOffset;
                 *pBaseReg = compLocallocUsed ? REG_SAVED_LOCALLOC_SP : REG_SPBASE;
             }
-            // Use ldr +/-imm8 encoding.
-            else if (offset >= -0x7C && offset <= ldrEncodeLimit)
+            // Use fp encoding.
```
Comment (briansull) [inline src/jit/compiler.hpp]: You might want to clarify this somehow.  I first read this comment "use fp encoding" as "use floating-point encoding"

File: src/jit/compiler.hpp [hunk 2]
```diff
@@ -2434,17 +2434,19 @@ inline
...
 
-            int spOffset       = fConservative ? compLclFrameSize : offset + codeGen->genSPtoFPdelta();
-            int actualOffset   = (spOffset + addrModeOffset);
-            int ldrEncodeLimit = (varTypeIsFloating(type) ? 0x3FC : 0xFFC);
-            // Use ldr sp imm encoding.
-            if (opts.MinOpts() || (actualOffset <= ldrEncodeLimit))
+            int spOffset           = fConservative ? compLclFrameSize : offset + codeGen->genSPtoFPdelta();
+            int actualOffset       = spOffset + addrModeOffset;
+            int encodingLimitUpper = varTypeIsFloating(type) ? 0x3FC : 0xFFF;
+            int encodingLimitLower = varTypeIsFloating(type) ? -0x3FC : -0xFF;
```
Comment (briansull) [inline src/jit/compiler.hpp]: Do we have good test coverage for this new range [0x-3FC .. orig] ?

**Pull request merged**